ci/ci/ci.py

```diff
@@ -121,13 +126,55 @@
 
     batch_client = request.app['batch_client']
     batches = batch_client.list_batches(
-        f'test=1 pr={pr_number}')
+        f'test=1 pr={pr.number}')
     batches = sorted([b async for b in batches], key=lambda b: b.id, reverse=True)
     page_context['history'] = [await b.status() for b in batches]
 
     return await render_template('ci', request, userdata, 'pr.html', page_context)
 
 
+async def retry_pr(wb, pr, request):
+    app = request.app
+    session = await aiohttp_session.get_session(request)
+
+    if pr.batch is None:
+        log.info('retry cannot be requested for PR #{pr.number} because it has no batch')
+        set_message(
+            session,
+            f'Retry cannot be requested for PR #{pr.number} because it has no batch.',
+            'error')
+        raise web.HTTPFound(
+            deploy_config.base_path('ci') +
+            f'/watched_branches/{wb.index}/pr/{pr.number}')
+    batch_id = pr.batch.id
+
+    dbpool = app['dbpool']
+    async with dbpool.acquire() as conn:
+        async with conn.cursor() as cursor:
+            await cursor.execute('INSERT INTO invalidated_batches (batch_id) VALUES (%s);', batch_id)
+
+    if pr.batch is not None and pr.batch.id == batch_id:
+        # no changes to the batch occured while we were talking to the database
+        pr.batch = None
+        await wb.notify_batch_changed(app)
```

Comment:
This part was a bit subtle for two reasons: the concurrency issue wrt pr.batch but also if I don't set the batch to None, then _update_batch will never check the invalidated batches.